Adds error when incorrect error codes are parsed from config file

mennoliefstingh opened the pull request:
This PR resolves #1689 by raising a warning when parsing a configuration file results in values for the --ignore or --extend-ignore options that cannot be valid error codes, i.e. they are not in a 'A123' format.

asottile commented on src/flake8/options/aggregator.py:
```diff
@@ -8,9 +8,11 @@
 from typing import Sequence
 
 from flake8.options import config
+from flake8.options.config import _find_config_file
```
private names should not be imported (also it's possible the values below don't come from those config files but instead from plugins -- so I don't think it's right to show the filename anyway)

mennoliefstingh replied:
good point, listing the invalid error code in the warning should already give the user enough information to know where it's coming from

asottile commented on src/flake8/options/aggregator.py:
```diff
@@ -43,6 +45,16 @@ def aggregate_options(
 
+        if config_name in ["ignore", "extend-ignore"]:
+            for error_code in value:
+                if not re.match("^[A-Z][0-9]{3}$", error_code):
```
this should reuse the same pattern that's used for validating entrypoints, notably they can be any error prefix here

asottile replied:
I don't like repeating myself but that's not right -- it should reuse the exact same pattern that's used for validating entrypoints

there should be no commas or spaces when it checks it

mennoliefstingh replied:
Ooh wait you're referring to `VALID_CODE = re.compile("^[A-Z]{1,3}[0-9]{0,3}$", re.ASCII)`. 
I missed that initially, sorry.

asottile commented on src/flake8/options/aggregator.py:
```diff
@@ -43,6 +45,16 @@ def aggregate_options(
             assert isinstance(dest_val, str)
             dest_name = dest_val
 
+        if config_name in ["ignore", "extend-ignore"]:
```
should use a set containment rather than list containment here -- also I think this might be too late in the process -- would prefer this get validated during argument / config parsing if possible

mennoliefstingh replied:
Also a good point, there is an earlier opportunity to check this immediately after option values have been normalized to a list of strings in config.py

asottile reviewed (commented):
will need a test as well

I think this should probably _error_ rather than warn since the configuration is just incorrect

asottile commented:
> Would you say instead of logging a warning we'd want to raise an exception then? E.g. like passing a non-existent config file raises an ExecutionError:
> 
> ```
> cfg = configparser.RawConfigParser()
>     if config is not None:
>         if not cfg.read(config, encoding="UTF-8"):
>             raise exceptions.ExecutionError(
>                 f"The specified config file does not exist: {config}"
>             )
> ```

ideally it would be early enough in the validation that it would raise the argument parsing error rather than `ExecutionError`

mennoliefstingh commented:
All right, fixes in these last few commits: 

- Updated the regex pattern to the one used for entry point validation
- Moved argument validation further up in the process, i.e. immediately after the option values are parsed to a list
- Invalid codes now throw a ParsingError instead of logging a warning
- Wrote a unit test that tests with example invalid input (as in issue #1689) for both ignore and extend-ignore

mennoliefstingh commented:
whoops except for one small mistake, let me fix it right quick

asottile commented on src/flake8/options/config.py:
```diff
@@ -120,7 +124,20 @@ def parse_config(
     return config_dict
+
+
+VALID_CODE = re.compile("^[A-Z]{1,3}[0-9]{0,3}$", re.ASCII)
```
you've got this defined in a couple places (3?) -- probably want to bring them all together

mennoliefstingh replied:
Good idea! I imported the constant from finder.py now -- I considered moving it upwards into utils.py but decided against it with the contribution guidelines in mind (no extraneous refactoring), but you could make a case that it's better to import downwards (both import from utils.py) than sideways as we do now. What are your thoughts on this?

asottile commented on src/flake8/options/config.py:
```diff
@@ -120,6 +121,16 @@ def parse_config(
+        if option_name in {"ignore", "extend-ignore"}:
+            for error_code in final_value:
+                if not VALID_CODE.match(error_code):
+                    raise configparser.ParsingError(
```
I don't think this is the right type -- the others raise `ValueError` so we should probably do similarly?

mennoliefstingh replied:
e.g. supplying an undecodable config file throws a ParsingError (```Raised when a configuration file does not follow legal syntax.```), but I agree that in this context it makes more sense to raise a ValueError since it'll also be raised if a user supplies an invalid code in otherwise valid syntax.

asottile commented on src/flake8/options/config.py:
```diff
@@ -96,7 +97,7 @@ def parse_config(
     cfg: configparser.RawConfigParser,
     cfg_dir: str,
 ) -> dict[str, Any]:
-    """Parse and normalize the typed configuration options."""
+    """Parse and normalize the typed configuration options"""
```
not sure why this changed?

asottile commented on src/flake8/options/config.py:
```diff
@@ -8,6 +8,7 @@
 
 from flake8 import exceptions
 from flake8.options.manager import OptionManager
+from flake8.plugins.finder import VALID_CODE
```
let's put this into flake8.defaults and call it `VALID_CODE_PREFIX`

asottile commented on tests/unit/test_options_config.py:
```diff
@@ -218,3 +218,15 @@ def test_parse_config_ignores_unknowns(tmp_path, opt_manager, caplog):
+        cfg, _ = config.load_config("setup.cfg", [], isolated=False)
+
+    with pytest.raises(configparser.ParsingError):
+        _ = config.parse_config(opt_manager, cfg, tmpdir)
```
this assignment does nothing

asottile commented on tests/unit/test_options_config.py:
```diff
@@ -218,3 +218,15 @@ def test_parse_config_ignores_unknowns(tmp_path, opt_manager, caplog):
+    with tmpdir.as_cwd():
+        cfg, _ = config.load_config("setup.cfg", [], isolated=False)
+
+    with pytest.raises(configparser.ParsingError):
```
if you're raising a specific exception with a specific message that should be tested:

```python
with pytest.raises(...) as excinfo:
    ...
msg, = excinfo.value.args
assert msg == ...
```

asottile commented on tests/unit/test_options_config.py:
```diff
@@ -218,3 +219,25 @@ def test_parse_config_ignores_unknowns(tmp_path, opt_manager, caplog):
+
+@pytest.mark.parametrize("option", ("ignore", "extend-ignore"))
+def test_invalid_ignore_codes_raise_error(tmpdir, opt_manager, option):
+    invalid_option = "//comment"
```
DRY here is bad -- it's much harder to understand what's going on in the test -- https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html

mennoliefstingh replied:
Cool, never thought about the fact that KISS is more important than DRY for testing. Learning a lot here 😄 really appreciate the comments.

I removed the `invalid_option` variable but did keep the others as they're imported from elsewhere or originate from the `mark.parametrize` decorator.

mennoliefstingh commented on tests/unit/plugins/finder_test.py:
```diff
@@ -43,7 +44,7 @@ def _loaded(plugin=None, obj=None, parameters=None):
     ),
 )
 def test_valid_plugin_prefixes(s):
-    assert finder.VALID_CODE.match(s)
+    assert VALID_CODE_PREFIX.match(s)
```
Good point, I moved them to a new `test_error_code_validation()` module.

asottile commented on tests/unit/test_error_code_validation.py:
```diff
@@ -0,0 +1,36 @@
+from __future__ import annotations
```
this module should be named after the module it tests

asottile commented on tests/unit/test_options_config.py:
```diff
@@ -218,3 +219,21 @@ def test_parse_config_ignores_unknowns(tmp_path, opt_manager, caplog):
+        config.parse_config(opt_manager, cfg, tmpdir)
+
+    expected = (
+        f"Error code //comment "
+        f"supplied to '{option}' option "
+        f"does not match {VALID_CODE_PREFIX.pattern}"
```
hmm maybe my feedback didn't land -- there shouldn't be interpolation in tests

asottile commented on tests/unit/test_options_config.py:
```diff
@@ -218,3 +218,35 @@ def test_parse_config_ignores_unknowns(tmp_path, opt_manager, caplog):
+        "Error code //comment supplied to 'ignore' option "
+        "does not match ^[A-Z]{1,3}[0-9]{0,3}$"
+    )
+    msg = excinfo.value.args[0]
+    assert msg == (expected)
```
why didn't you use the code I wrote for you in the comment?  this assertion is weaker than the one I wrote for you as it allows an exception with more arguments to pass silently

also these parens do nothing

mennoliefstingh replied:
Hi, sorry about being quiet for a while but I caught a nasty covid case. Fixed this assertion now!

asottile reviewed (commented):
can you show the commandline experience when this fails?

mennoliefstingh commented:
> can you show the commandline experience when this fails?

Error being raised with a dodgy .flake8 file: 

`flake8
Traceback (most recent call last):
  File "/usr/local/bin/flake8", line 33, in <module>
    sys.exit(load_entry_point('flake8', 'console_scripts', 'flake8')())
  File "/Users/menno/dev/flake8/src/flake8/main/cli.py", line 23, in main
    app.run(argv)
  File "/Users/menno/dev/flake8/src/flake8/main/application.py", line 333, in run
    self._run(argv)
  File "/Users/menno/dev/flake8/src/flake8/main/application.py", line 321, in _run
    self.initialize(argv)
  File "/Users/menno/dev/flake8/src/flake8/main/application.py", line 306, in initialize
    self.parse_configuration_and_cli(cfg, cfg_dir, remaining_args)
  File "/Users/menno/dev/flake8/src/flake8/main/application.py", line 148, in parse_configuration_and_cli
    self.options = aggregator.aggregate_options(
  File "/Users/menno/dev/flake8/src/flake8/options/aggregator.py", line 30, in aggregate_options
    parsed_config = config.parse_config(manager, cfg, cfg_dir)
  File "/Users/menno/dev/flake8/src/flake8/options/config.py", line 128, in parse_config
    raise ValueError(
ValueError: Error code // supplied to 'ignore' option does not match ^[A-Z]{1,3}[0-9]{0,3}$`

asottile reviewed (approved):
![](https://i.fluffy.cc/4HrdmGw7MHL3rcH3J0lCBBG7xhlxnWs7.gif)
